sdk/provisioning/Azure.Provisioning/src/ModuleConstruct.cs

```diff
@@ -221,17 +239,43 @@ private void WriteParameters(MemoryStream stream)
                 {
                     continue;
                 }
-                string defaultValue =
-                    parameter.DefaultValue is null ?
-                        string.Empty :
-                        parameter.IsExpression ? $" = {parameter.DefaultValue}" : $" = '{parameter.DefaultValue}'";
 
                 if (parameter.IsSecure)
+                {
                     stream.WriteLine($"@secure()");
+                }
 
                 stream.WriteLine($"@description('{parameter.Description}')");
-                stream.WriteLine($"param {parameter.Name} string{defaultValue}{Environment.NewLine}");
+                stream.WriteLine($"{GetParameterDeclaration(parameter)}{Environment.NewLine}");
+            }
+        }
+
+        private string GetParameterDeclaration(Parameter parameter)
+        {
+            string value;
+
+            if (parameter.DefaultValue is null)
+            {
+                value = string.Empty;
+            }
+            else if (parameter.IsExpression)
+            {
+                value = parameter.DefaultValue.ToString()!;
             }
+            else
+            {
+                value = parameter.Kind switch
+                {
+                    BicepKind.Bool => parameter.DefaultValue.ToString()!.ToLower(),
+                    BicepKind.Int => parameter.DefaultValue.ToString()!,
+                    BicepKind.Array => parameter.DefaultValue.ToString()!,
+                    BicepKind.Object => parameter.DefaultValue.ToString()!,
+                    BicepKind.String => $"'{parameter.DefaultValue}'",
+                    _ => throw new NotSupportedException("Invalid parameter kind.")
+                };
+            }
+
+            return $"param {parameter.Name} {parameter.Kind.ToString().ToLower()}{(!string.IsNullOrEmpty(value) ? $" = {value}" : string.Empty)}";
```

Comment:
Nested string interpolation is kind of hard to read - this might be easier to maintain if you handled the default value after the main declaration